File: src-tauri/src/commands/mantenimiento.rs
//! commands/mantenimiento.rs — Comandos Tauri del módulo de mantenimiento (thin wrappers)

use crate::core::error::ErrorPayload;
use crate::repositories::mantenimiento::{Mantenimiento, MantenimientoDatos};
use crate::services::mantenimiento::{AlertaKm, MantenimientoService, TotalesMantenimiento};
use crate::services::AppState;
use tauri::State;

use super::{conn, require_session};

type Cmd<T> = Result<T, ErrorPayload>;

/// Lista mantenimientos con filtros opcionales (búsqueda, placa o tipo)
#[tauri::command]
pub fn listar_mantenimientos(
    state: State<'_, AppState>,
    session_id: String,
    busqueda: Option<String>,
    placa: Option<String>,
    tipo: Option<String>,
) -> Cmd<Vec<Mantenimiento>> {
    require_session(&state, &session_id)?;
    let mut c = conn(&state)?;
    MantenimientoService::listar(&mut c, busqueda.as_deref(), placa.as_deref(), tipo.as_deref())
        .map_err(|e| e.to_payload())
}

/// Mantenimientos recientes
#[tauri::command]
pub fn mantenimientos_recientes(
    state: State<'_, AppState>,
    session_id: String,
    limit: Option<i64>,
) -> Cmd<Vec<Mantenimiento>> {
    require_session(&state, &session_id)?;
    let mut c = conn(&state)?;
    MantenimientoService::recientes(&mut c, limit.unwrap_or(10)).map_err(|e| e.to_payload())
}

/// Obtiene un mantenimiento por id
#[tauri::command]
pub fn obtener_mantenimiento(state: State<'_, AppState>, session_id: String, id: i64) -> Cmd<Mantenimiento> {
    require_session(&state, &session_id)?;
    let mut c = conn(&state)?;
    MantenimientoService::obtener(&mut c, id).map_err(|e| e.to_payload())
}

/// Crea un mantenimiento
#[tauri::command]
pub fn crear_mantenimiento(
    state: State<'_, AppState>,
    session_id: String,
    datos: MantenimientoDatos,
) -> Cmd<Mantenimiento> {
    require_session(&state, &session_id)?;
    let mut c = conn(&state)?;
    MantenimientoService::crear(&mut c, &state.config, datos).map_err(|e| e.to_payload())
}

/// Actualiza un mantenimiento por id
#[tauri::command]
pub fn actualizar_mantenimiento(
    state: State<'_, AppState>,
    session_id: String,
    id: i64,
    datos: MantenimientoDatos,
) -> Cmd<Mantenimiento> {
    require_session(&state, &session_id)?;
    let mut c = conn(&state)?;
    MantenimientoService::actualizar(&mut c, &state.config, id, datos).map_err(|e| e.to_payload())
}

/// Elimina un mantenimiento
#[tauri::command]
pub fn eliminar_mantenimiento(state: State<'_, AppState>, session_id: String, id: i64) -> Cmd<()> {
    require_session(&state, &session_id)?;
    let mut c = conn(&state)?;
    MantenimientoService::eliminar(&mut c, id).map_err(|e| e.to_payload())
}

/// Totales general, por placa y por tipo
#[tauri::command]
pub fn totales_mantenimiento(state: State<'_, AppState>, session_id: String) -> Cmd<TotalesMantenimiento> {
    require_session(&state, &session_id)?;
    let mut c = conn(&state)?;
    MantenimientoService::totales(&mut c).map_err(|e| e.to_payload())
}

/// Alertas por kilometraje (cambio de aceite y frenos próximos o vencidos)
#[tauri::command]
pub fn alertas_km_mantenimiento(state: State<'_, AppState>, session_id: String) -> Cmd<Vec<AlertaKm>> {
    require_session(&state, &session_id)?;
    let mut c = conn(&state)?;
    MantenimientoService::alertas_km(&mut c, &state.config).map_err(|e| e.to_payload())
}

File: src/lib/api/mantenimiento.ts
// mantenimiento.ts — Comandos y tipos para mantenimiento de vehículos
import { invokeCmd } from './base';

/** Registro de mantenimiento (repositories/mantenimiento.rs) */
export interface Mantenimiento {
	id: number;
	placa: string;
	vehiculo: string;
	tipo: string;
	fecha: string;
	descripcion: string | null;
	observaciones: string | null;
	costo: string;
	kmProximoCambioAceite: number | null;
	total: string;
	createdAt: string | null;
	updatedAt: string | null;
}

/** Datos de entrada para crear/actualizar un mantenimiento */
export interface MantenimientoDatos {
	placa: string;
	tipo: string;
	fecha: string;
	descripcion?: string;
	observaciones?: string;
	costo: string;
	kmProximoCambioAceite?: number | null;
}

/** Total por placa o tipo */
export interface TotalMantenimiento {
	clave: string;
	total: string;
}

/** Resumen de totales de mantenimiento (services/mantenimiento.rs) */
export interface TotalesMantenimiento {
	totalGeneral: string;
	porPlaca: TotalMantenimiento[];
	porTipo: TotalMantenimiento[];
}

/** Alerta por kilometraje (cambio de aceite o frenos próximo/vencido) */
export interface AlertaKm {
	placa: string;
	marca: string;
	modelo: string;
	tipo: string;
	kmActual: number;
	kmProximo: number;
	kmRestante: number;
	critica: boolean;
}

export const mantenimientoApi = {
	listar: (sessionId: string, busqueda?: string, placa?: string, tipo?: string) =>
		invokeCmd<Mantenimiento[]>('listar_mantenimientos', {
			sessionId,
			busqueda: busqueda || null,
			placa: placa || null,
			tipo: tipo || null
		}),
	recientes: (sessionId: string, limit?: number) =>
		invokeCmd<Mantenimiento[]>('mantenimientos_recientes', { sessionId, limit: limit ?? null }),
	obtener: (sessionId: string, id: number) =>
		invokeCmd<Mantenimiento>('obtener_mantenimiento', { sessionId, id }),
	crear: (sessionId: string, datos: MantenimientoDatos) =>
		invokeCmd<Mantenimiento>('crear_mantenimiento', { sessionId, datos }),
	actualizar: (sessionId: string, id: number, datos: MantenimientoDatos) =>
		invokeCmd<Mantenimiento>('actualizar_mantenimiento', { sessionId, id, datos }),
	eliminar: (sessionId: string, id: number) =>
		invokeCmd<void>('eliminar_mantenimiento', { sessionId, id }),
	totales: (sessionId: string) =>
		invokeCmd<TotalesMantenimiento>('totales_mantenimiento', { sessionId }),
	alertasKm: (sessionId: string) => invokeCmd<AlertaKm[]>('alertas_km_mantenimiento', { sessionId })
};
